### PR TITLE
[NPU] Add Qwen3.5-35B-A3B support

### LINKED ISSUE #389: [NPU] Add Qwen3.5-35B-A3B support
# Summary

This issue documents the end-to-end deployment and training workflow for running **Qwen3.5-35B-A3B** GRPO training with Vime on Ascend NPU (Atlas A3), intended as a reference for the community.

| Item | Version | Commit |
|---|---|---|
| Ascend HDK | `25.5.1.B030` | `N/A` |
| CANN | `9.0.0` | `N/A` |
| python | `3.12.13` | `N/A` |
| **torch_npu** | **`2.10.0.post2`** | **`N/A`** |
| triton-ascend | `3.2.1` | `N/A` |
| vllm | `0.22.1+empty` | `0decac0` |
| vllm-ascend | `0.22.1rc1` | `5f6faa0c` |
| **Megatron-Bridge** | **`N/A`** | **`7f0fb345`** |
| Megatron-LM | `0.16.0rc0` | `1dcf0dafa` |
| MegatronAdaptor | `0.1.0` | `15582ad` |
| TransformerEngineNPU | `2.13.0` | `d743c83` |
| **flash-linear-attention-npu** | **`v26.6.0`** | **`14c2c92`** |

# Usage
> ⚠️ Note: Before proceeding, you must update the following paths to match your server environment:
- Docker volume mappings in the container execution command
- Local storage directories designated for model and dataset downloads
- Internal paths and configurations defined within the training script

## 1. Setup

```bash
git clone -b ascend https://github.com/vllm-project/vime.git
cd vime
docker build -f docker/Dockerfile.npu -t vime-ascend:latest .
```

```bash
export IMAGE=vime-ascend:latest

docker run -d --name vime-npu -it --net=host --shm-size=1024g \
    --privileged=true \
    --cap-add=SYS_PTRACE \
    --device=/dev/davinci_manager \
    --device=/dev/hisi_hdc \
    --device=/dev/devmm_svm \
    -v /usr/local/Ascend/driver:/usr/local/Ascend/driver \
    -v /usr/local/dcmi:/usr/local/dcmi \
    -v /usr/local/bin/npu-smi:/usr/local/bin/npu-smi \
    -v /usr/local/sbin:/usr/local/sbin \
    -v /home:/home \
    -v /mnt:/mnt \
    -v /tmp:/tmp \
    -v /data:/data \
    -v /usr/share/zoneinfo/Asia/Shanghai:/etc/localtime \
    $IMAGE

docker exec -it vime-npu bash
```

## 2. Download

```bash
# (1) Model
hf download Qwen/Qwen3.5-35B-A3B --local-dir /path/to/Qwen3.5-35B-A3B

# (2) Data
hf download --repo-type dataset zhuzilin/dapo-math-17k --local-dir /path/to/dapo-math-17k
hf download --repo-type dataset zhuzilin/aime-2024  --local-dir /path/to/aime-2024
```

## 3. Training ([run-qwen3.5-35B-A3B-npu.sh](https://github.com/vllm-project/vime/blob/ascend/scripts/run-qwen3.5-35B-A3B-npu.sh))

```bash
cd /root/vime
bash scripts/run-qwen3.5-35B-A3B-npu.sh
```

# Results
reward
diff
eval

# Proposed Follow-ups
- [ ] **Integrate fla-npu native `causal_conv1d` operator**
Currently, the causal_conv1d in the Qwen3.5-35B-A3B GDN module uses the Triton implementation (megatron.core.ssm.triton.causal_conv1d).

- [ ] **Enable vLLM graph mode on Ascend (remove `--vllm-enforce-eager`)**
Currently, the rollout engine forces eager mode by hardcoding --vllm-enforce-eager in scripts/run-qwen3.5-35B-A3B-npu.sh due to unsupported NPU graph capture, disabling vLLM's graph-based decode acceleration.

- [ ] **Support dynamic batching for the GDN path (`--use-dynamic-batch-size`)**
Lacking Megatron GDN packing support, the training script is forced to use --micro-batch-size 1 with heavy padding, disabling dynamic batching.

### PR DESCRIPTION
**Pull request by Fulin-Gao**
# [NPU] Add Qwen3.5-35B-A3B RL training example on Ascend NPU

## Overview

This PR adds a **Qwen3.5-35B-A3B** GRPO reinforcement-learning training example to vime's Ascend NPU backend. Qwen3.5-35B-A3B is a **hybrid-architecture** model: its attention path uses linear attention (**Gated Delta Net**, GDN), and its FFN path is a 256-expert MoE. Enabling this on NPU requires two new pathways to work end-to-end:

1. **GDN operator path** — NPU-native `chunk_gated_delta_rule` fwd/bwd kernels provided by `fla-npu` (flash-linear-attention-npu);
2. **MoE training + inference path** — Megatron training at TP=2 / EP=8 with vLLM rollout weight sync.

The example ships a reproducible RL-stage script, bilingual docs, a CI smoke test, plus the Docker image changes and Megatron / Megatron-Bridge patch upgrades required to run this model on NPU.

## Files

| File | Description |
|------|-------------|
| `scripts/run-qwen3.5-35B-A3B-npu.sh` | Main launch script: GRPO RL training of Qwen3.5-35B-A3B on 16 NPUs, with Ray isolation, NPU env vars, and parallelism/optimizer/vLLM/eval args |
| `tests/test_qwen3.5_35B_A3B_npu.py` | CI smoke test: smoke-scaled rollout (num-rollout=2, batch=4, n-samples=4), TP=2/EP=8, `--ci-test`, auto-downloads model and data |
| `docs/en/examples/qwen3.5-35B-A3B-npu.md` | English doc: container startup, fla-npu build/install, model/data download, run training |
| `docs/zh/examples/qwen3.5-35B-A3B-npu.md` | Chinese doc (same) |
| `.buildkite/npu_suites.py` | Register `test_qwen3.5_35B_A3B_npu.py` in the `npu-16` CI suite |
| `docker/Dockerfile.npu` | Install fla-npu in-image, pin `torch_npu==2.10.0.post2`, bump Megatron-Bridge commit, add `fla_npu` to the import smoke check |
| `docker/npu_patch/megatron.patch` | Add the full `megatron/core/ssm/` GDN stack (NPU `fla_npu` op dispatch + triton kernels) plus several NPU adaptations |
| `docker/npu_patch/megatron-bridge.patch` | Param mapping supports `MindSpeedTELayerNormColumnParallelLinear`; Qwen3-VL swaps NPU-safe `nvtx_range_*` for `torch.cuda.nvtx`; etc. |
| `docker/npu_patch/README.md` | Sync Megatron-Bridge commit and `torch-npu==2.10.0.post2` version |

## Results
reward
diff
eval

Refer to #389 for steps, closes #389.

### CODE REVIEW

File: scripts/run-qwen3.5-35B-A3B-npu.sh [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+	  --dashboard-port="${RAY_DASHBOARD_PORT}" \
+	  --dashboard-agent-listen-port="${RAY_AGENT_PORT}" \
+	  --node-ip-address 127.0.0.1 \
+	  --num-gpus 0 \
+	  --resources "{\"NPU\": $NUM_NPUS}" \
+	  --disable-usage-stats \
+	  --dashboard-host=0.0.0.0
+
+# Build the runtime environment JSON with proper variable substitution
+RUNTIME_ENV_JSON=$(cat << 'EOF'
```
Comment (CalvinXKY) [inline scripts/run-qwen3.5-35B-A3B-npu.sh:166]: Ray job workers override the correct shell `PYTHONPATH` from L37, so imports will fail.

File: scripts/run-qwen3.5-35B-A3B-npu.sh [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  "env_vars": {
+    "PYTHONPATH": "${VIME_DIR}:/root/Megatron-LM:/vllm-workspace/vllm:/vllm-workspace/vllm-ascend:/root/Megatron-Bridge/src:/root/mbridge:/root/MegatronAdaptor:/root/TransformerEngineNPU:/usr/local/Ascend/ascend-toolkit/latest/python/site-packages",
+    "CUDA_DEVICE_MAX_CONNECTIONS": "1",
+    "HCCL_HOST_SOCKET_PORT_RANGE": "60000-60050",
+    "HCCL_NPU_SOCKET_PORT_RANGE": "61000-61050",
+    "HCCL_CONNECT_TIMEOUT": "7200",
+    "PYTORCH_NPU_ALLOC_CONF": "expandable_segments:False",
+    "VLLM_DISABLE_COMPILE_CACHE": "1",
+    "TRANSFORMERS_VERBOSITY": "error",
+    "RUST_LOG": "vllm_router_rs=warn",
```
Comment (CalvinXKY) [inline scripts/run-qwen3.5-35B-A3B-npu.sh:177]: The runtime_env block (L167–179) omits NPU vars exported above: `HCCL_DETERMINISTIC` (L45), `VLLM_ASCEND_ENABLE_NZ` (L46), `ATB_MATMUL_SHUFFLE_K_ENABLE` (L48), `ATB_LLM_LCOC_ENABLE` (L49), `TASK_QUEUE_ENABLE` (L50), `RAY_EXPERIMENTAL_NOSET_ASCEND_RT_VISIBLE_DEVICES` (L52). Job workers may not get all of these once `--runtime-env-json` is applied.

File: docs/en/examples/qwen3.5-35B-A3B-npu.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    -v /mnt:/mnt \
+    -v /tmp:/tmp \
+    -v /data:/data \
+    -v /usr/share/zoneinfo/Asia/Shanghai:/etc/localtime \
+    $IMAGE
+
+docker exec -it vime-npu bash
+```
+
+## 2. Install fla-npu
```
Comment (CalvinXKY) [inline docs/en/examples/qwen3.5-35B-A3B-npu.md]: Asks users to manually install fla-npu, but this PR already bakes it into `docker/Dockerfile.npu`.

File: docs/en/examples/qwen3.5-35B-A3B-npu.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+```bash
+# (1) Model
+hf download Qwen/Qwen3.5-35B-A3B --local-dir /path/to/Qwen3.5-35B-A3B
+
+# (2) Data
+hf download --repo-type dataset zhuzilin/dapo-math-17k --local-dir /path/to/dapo-math-17k
+hf download --repo-type dataset zhuzilin/aime-2024  --local-dir /path/to/aime-2024
+```
+
+## 4. Run Training
```
Comment (CalvinXKY) [inline docs/en/examples/qwen3.5-35B-A3B-npu.md]: `/path/to/...` checkpoint/data paths (L63–66, L73, L90) must be edited before running, but docs §4 (L75–79) only says to run the script. 

Add an explicit “edit paths first” step, or use a `DATA_ROOT` variable like `run-qwen3-30B-A3B-npu.sh`  ？

File: docs/en/examples/qwen3.5-35B-A3B-npu.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
+# Training Qwen3.5-35B-A3B on Ascend NPU
```
Comment (CalvinXKY) [inline docs/en/examples/qwen3.5-35B-A3B-npu.md]: Per the existing convention, adding the model .sh script is sufficient for a model; please avoid adding extra standalone intro/docs under examples/.

File: scripts/run-qwen3.5-35B-A3B-npu.sh (hunk 1 above)
Comment (Fulin-Gao) [reply scripts/run-qwen3.5-35B-A3B-npu.sh:166]: No import issues.

File: scripts/run-qwen3.5-35B-A3B-npu.sh (hunk 2 above)
Comment (Fulin-Gao) [reply scripts/run-qwen3.5-35B-A3B-npu.sh:177]: Got it. Thanks for confirming that the configuration has been verified through multiple tests and the unnecessary variables were cleaned up.

File: docs/en/examples/qwen3.5-35B-A3B-npu.md (hunk 4 above)
Comment (Fulin-Gao) [reply docs/en/examples/qwen3.5-35B-A3B-npu.md]: Markdown files removed; no longer applicable.

File: docs/en/examples/qwen3.5-35B-A3B-npu.md (hunk 5 above)
Comment (Fulin-Gao) [reply docs/en/examples/qwen3.5-35B-A3B-npu.md]: Markdown files removed; no longer applicable.

File: docs/en/examples/qwen3.5-35B-A3B-npu.md (hunk 3 above)
Comment (Fulin-Gao) [reply docs/en/examples/qwen3.5-35B-A3B-npu.md]: Markdown files removed; no longer applicable.

**Review by CalvinXKY — APPROVED** [review]

**Comment by justforyou16007** [comment]
用A2*2节点跑run-qwen3.5-35B-A3B-npu.sh会提示RPC sample token timeout，增加日志后提示卡点在 GDN 线性注意力的 chunked-prefill 路径，而且这个错误是在安装fla-npu之后出现的，不安装fla-npu可以正常rollout但是训练会报算子未发现问题
